cpp/src/arrow/compute/exec/expression.cc

```diff
@@ -61,13 +61,22 @@ Expression call(std::string function, std::vector<Expression> arguments,
   call.function_name = std::move(function);
   call.arguments = std::move(arguments);
   call.options = std::move(options);
+
+  call.hash = std::hash<std::string>{}(call.function_name);
+  for (const auto& arg : call.arguments) {
+    call.hash ^= arg.hash();
```

Comment:
Note that simple XORing is suboptimal (`a,b` will produce the same hash as `b,a`, for example).
May want to use something like `hash_combine` from `arrow/util/hash_util.h`.

Reply:
thanks, will replace